File: Data Structures & Algorithms/surrounded-regions/submission-57.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        ROWS, COLS = len(board), len(board[0])
        q = deque()
        connected = set()
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        for r in range(ROWS): 
            if board[r][0] == "O": 
                q.append((r, 0))
                connected.add((r, 0))
            if board[r][COLS - 1] == "O": 
                q.append((r, COLS - 1))
                connected.add((r, COLS - 1))
        for c in range(COLS): 
            if board[0][c] == "O": 
                q.append((0, c))
                connected.add((0, c))
            if board[ROWS - 1][c] == "O": 
                q.append((ROWS - 1, c))
                connected.add((ROWS - 1, c))
        
        while q: 
            r, c = q.popleft()
            for dr, dc in directions: 
                row, col = r + dr, c + dc
                if (row in range(ROWS) and col in range(COLS) and board[row][col] == "O"
                    and (row, col) not in connected): 
                    q.append((row, col))
                    connected.add((row, col))
        
        for r in range(ROWS): 
            for c in range(COLS): 
                if board[r][c] == "O" and (r, c) not in connected: 
                    board[r][c] = "X"

File: Data Structures & Algorithms/search-for-word/submission-57.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int ROWS = board.length, COLS = board[0].length;
        boolean[][] visited = new boolean[ROWS][COLS];

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (dfs(r, c, 0, board, word, visited) == true) return true;
            }
        }
        return false;
    }
    public boolean dfs(int r, int c, int i, char[][] board, String word, boolean[][] visited) {
        if (i == word.length()) return true;

        if (r < 0 || r == board.length || c < 0 || c == board[0].length) return false;
        if (board[r][c] != word.charAt(i) || visited[r][c] == true) return false;

        visited[r][c] = true;
        boolean result = (dfs(r + 1, c, i + 1, board, word, visited) ||
                          dfs(r - 1, c, i + 1, board, word, visited) ||
                          dfs(r, c + 1, i + 1, board, word, visited) ||
                          dfs(r, c - 1, i + 1, board, word, visited));
        visited[r][c] = false;
        return result;
    }
}
